File: src/python/solve.py
import sudoku
import math
import re

input = open(0).read().splitlines()

puzzle = []
for line in input:
    cells = line.split(' ')
    row = [int(x) if re.match(r'^-?\d+$', x) != None else 0 for x in cells]
    puzzle.append(row)

NN = len(puzzle)
N = math.isqrt(len(puzzle))
if N * N != NN:
    print('Invalid input: grid size does not have integer square root')
    exit()

for idx, row in enumerate(puzzle):
    if len(row) != NN:
        print('Invalid input: size of row with idx {} does not match columns count'.format(idx))
        exit()

for row in puzzle:
    for value in row:
        if value < 0 or value > NN:
            print('Invalid input: incorrect value {}'.format(value))
            exit()

solution = sudoku.get_random_solution(puzzle, N)
if solution == None:
    print("Solution not found")
    exit()

for row in solution:    
    print(' '.join([str(x) for x in row]))

File: src/c/solve.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "sudoku.h"

int main(int argc, char* argv[]) {
    reseed_rand(time(NULL));
    int puzzle[SUDOKU_NN][SUDOKU_NN];

    for(int i = 0; i < SUDOKU_NN; ++i) {
        for(int j = 0; j < SUDOKU_NN; ++j) {
            if(scanf("%d", &puzzle[i][j]) != 1 || puzzle[i][j] < 0 || puzzle[i][j] > SUDOKU_NN) {
                printf("Invalid input value. Only digits are acceptable.\n");
                return -1;
            }
        }
    }

    int (*solution)[SUDOKU_NN] = (int(*)[SUDOKU_NN]) get_random_solution(puzzle);
    if(solution == NULL) {
        printf("No solution found.\n");
        return 0;
    }

    for(int i = 0; i < SUDOKU_NN; ++i) {
        for(int j = 0; j < SUDOKU_NN; ++j) {
            printf("%d ", solution[i][j]);
        }
        printf("\n");
    }
    free(solution);
}